wordpress-mobile/WordPress-iOS
# Reader: Cache attributed string building to avoid crashes and performance issues

Fixes #21461

Building NSAttributedString from HTML within tableView cell configuration methods triggers nested table view reloads that cause inconsistency between tableView and resultController states and lead to a crash.

More details: https://github.com/wordpress-mobile/WordPress-iOS/issues/21461#issuecomment-1840435759

### Solution

A caching mechanism was created and still exists within ReaderCommentsViewController but was abandoned for no clear reason and not used. Building attributed string from html inside cell configuration creates performance issues and also triggers CoreData and tableView reloads that cause crashes. NSAttributedString should be built outside tableView cellForRowForIndexPath and cached for performance.

I pass cached attributed string to a cell that was already being cached in `updateCachedContent` of `ReaderCommentsViewController`  https://github.com/wordpress-mobile/WordPress-iOS/pull/11582 

## To test:

1. Select a comment from notifications
2. Change to "Pending"
3. Open comment
4. Come back
5. Change to "Approved"
6. Open comment
7. No crash should happen and view should scroll to comment

Video: https://github.com/wordpress-mobile/WordPress-iOS/issues/21461#issuecomment-1840435759

## Regression Notes
1. Potential unintended areas of impact

I want to make sure that Reader comments loading is not breaking in any way. The caching mechanism looks to be quite safe, since it creates attributed text if it doesn't exist yet, so there shouldn't be cases where the comment doesn't appear.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing of Reader comments.

3. What automated tests I added (or what prevented me from doing so)

None

PR submission checklist:


UI Changes testing checklist:
